File: archaeology/200901-source-without-versioning/wengophone-ng/pywengophone/src/chat/emoticons/EmoticonsWidget.py
# -*- coding: utf-8 -*-
#
# WengoPhone, a voice over Internet phone
# Copyright (C) 2004-2007 Wengo
#
# This program is free software you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with self program if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
#
# @author Aurelien Gateau
# @author Ovidiu Ciule

from EmoticonsManager import *
from EmoticonButton import *
from Emoticon import *
from util.SafeConnect import *
from PyQt4.QtCore import *
from PyQt4.QtGui import *

import logging

log = logging.getLogger("PyWengoPhone.EmoticonsWidget")

class EmoticonsWidget(QWidget):
    """
    Emoticons widget
    """
    def __init__(self, parent, protocol = "default"):
        QWidget.__init__(self, parent)
        self.initLayout()
	self.__parent = parent
        self.initButtons(protocol)

    def emoticonClickedSlot(self, emoticon):
        """
        Centralizes the emoticon clicked signals
        """
	log.info("emoticonClicked"+emoticon.getText()[0])
	self.__parent.close()
	self.emit(SIGNAL("emoticonClicked"), emoticon)

    def initLayout(self):
        self.__layout = QGridLayout(self)
        self.__layout.setMargin(0)
        self.__layout.setSpacing(0)
 
    def initButtons(self, protocol):
        """
        Loads the emoticons of the protocol
        """
        for child in self.__layout.children():
            self.__layout.removeWidget(child)
            del(child)
        self.__buttonX = 0
        self.__buttonY = 0
        emoticonsManager = EmoticonsManager.getInstance()
        emoticonList = emoticonsManager.getEmoticonList(protocol)
        for emoticon in emoticonList: 
            self.addButton(emoticon)
        
    def addButton(self, emoticon):
        """
        Adds a button; uses a grid 'layout'
        """
        if (self.__buttonX == 10):
                self.__buttonX = 0
                self.__buttonY += 1
        
        button = EmoticonButton(self, emoticon)
        self.__layout.addWidget(button, self.__buttonY, self.__buttonX)
        connect(button, SIGNAL("buttonClicked"), self.emoticonClickedSlot)
        self.__buttonX += 1

File: archaeology/200901-source-without-versioning/wengophone-1.0/gui/chat/EmoticonsWidget.cpp
/*
* WengoPhone, a voice over Internet phone
* Copyright (C) 2004-2005  Wengo
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "ChatPageWidget.h"
#include "EmoticonsWidget.h"
#include "MyWidgetFactory.h"

#include <qwidget.h>
#include <qpushbutton.h>
#include <qlayout.h>
#include <qiconset.h>
#include <qtooltip.h>
#include <qurl.h>
#include <qobjectlist.h>
#include <qcursor.h>
#include <qapplication.h>
#include <qdir.h>

#include <iostream>
using namespace std;

EmoticonsWidget::EmoticonsWidget(QObject * parent, bool useTextLabel, int nbcol)
	: QObject(parent) {
	_emoticonsWidget = MyWidgetFactory::create("EmoticonsWidgetForm.ui", this, 0);
#ifdef WIN32
	_emoticonsWidget->recreate(0, Qt::WStyle_Splash, _emoticonsWidget->pos());
#endif
	_layout = (QGridLayout*)_emoticonsWidget->layout();
	_nbcol = nbcol;
	_useTextLabel = useTextLabel;
}

void EmoticonsWidget::setIconset(Iconset iconset) {
	_iconset.clear();
	_iconset = iconset;
	update();
}

void EmoticonsWidget::setIconset(const QString & /*name*/) {
	Iconset iconset;
	iconset.load(qApp->applicationDirPath() + QDir::separator() + "emoticons" + QDir::separator() + "default");
	setIconset(iconset);
}

void EmoticonsWidget::update() {
	erase();
	QPtrListIterator<Icon> it = _iconset.iterator();
	Icon *icon;
	QPixmap pix;
	int count = 0;
	int row = 0;
	while ( (icon = it.current()) != 0 ) {
		++it;
		QPushButton * b = new QPushButton(_emoticonsWidget);
		connect(b, SIGNAL(clicked()), SLOT(clicked()));
		//b->setUsesTextLabel(true);
		b->setMaximumSize(50, 50);
		b->setFlat(true);
		QString temp = regExp2HumanReadable(icon->regExp());
		//b->setTextPosition(QToolButton::BesideIcon);
		b->setName(getFirstShortcut(icon->regExp()));
		if(_useTextLabel) {
			b->setText(getFirstShortcut(QRegExp::escape(temp)));
		}
		b->setIconSet(QIconSet(icon->pixmap()));
		QToolTip::add(b, regExp2HumanReadable(icon->regExp()));
		b->show();
		if( count >= _nbcol) {
			count = 0;
			row++;
		}
		_layout->addWidget((QWidget*)b, row, count);
		count++;
	}
	_emoticonsWidget->adjustSize();
}

QString EmoticonsWidget::regExp2HumanReadable(QRegExp regexp) {
	QString temp = regexp.pattern();
	temp.replace("\\|", "pipereplacement");
	temp.replace("\\\\", "backslashreplacement");
	temp.remove("\\");
	temp.replace("|", tr(" or "));
	temp.replace("pipereplacement", "|");
	temp.replace("backslashreplacement", "\\");
	return temp;
}

QString EmoticonsWidget::getFirstShortcut(QRegExp regexp) {
	QString temp = regExp2HumanReadable(regexp);
	int index = temp.find(tr(" or"));
	temp.remove(index, temp.length() - index);
	return temp;
}

void EmoticonsWidget::erase() {
	const QObjectList * children = ((QObject*)_emoticonsWidget)->queryList("QToolButton");
	for(QObjectList::iterator it = children->begin(); it != children->end(); it++) {
		disconnect(SIGNAL(clicked()), this, SLOT(clicked()));
		delete(*it);
	}
}

void EmoticonsWidget::clicked() {
	QPushButton * button = (QPushButton*)_emoticonsWidget->childAt(_emoticonsWidget->mapFromGlobal(_emoticonsWidget->cursor().pos()));
	if(button) {
		((ChatPageWidget*)parent())->insertImage(button->name());
		((ChatPageWidget*)parent())->giveFocusToTextEdit();
	}
}

QMimeSourceFactory *EmoticonsWidget::getMimeSourceFactory() {
	return _iconset.createMimeSourceFactory();
}
